Create chats (2-users)

blong1996 opened the pull request:
This PR allows a user to create a chat with another user from within the application by clicking the "+" icon on the `ConversationFragment` or by clicking a User from the `UserFragment`.
This PR also adds toolbars (title bars) to each fragment and activity.

blong1996 committed:
Update Chats Fragment Toolbar

- Added toolbar to chats fragment with search and Add Chat image buttons
- Removed toolbar from Tabs Activity

blong1996 committed:
Make Create Chat Activity UI

- complete UI for creating new chats
- add default group avatar image to drawable directory

blong1996 committed:
Create Add Participants Activity

- rename create chat activity to create group

blong1996 committed:
Merge latest version of dev branch

- Merge PopulateUsers PR

blong1996 committed:
Merge branch 'dev' of https://github.com/ymittal/codeu_project_2017 into create-chats

blong1996 committed:
Add Toolbars to all fragments

blong1996 committed:
Fix id names and implement strings

blong1996 committed:
Merge dev branch

blong1996 committed:
- Remove TabActivity toolbar
- Create “Create Conversation” Activity
- Create Conversations Presenter
- Create “Create Conversation” View

blong1996 committed:
Populate Create Conversation Activity with users

blong1996 committed:
Successfully create new one-to-one chat

- Create “CreateConversationInteractor”  & “CreateConversationPresenter”
- Implement creating convo by clicking user object in userListAdapter
(Can be done from “CreateConversationActivity” or “UserFragment”

ymittal commented:
Please use Reformat Code shortcut for the Java and XML files changed in this PR. There's minor indentation and spacing fixes which will be fixed automatically.

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/presenter/ConversationsPresenter.java:
```diff
@@ -4,6 +4,7 @@
 
 import com.google.codeu.chatme.model.Conversation;
 import com.google.codeu.chatme.model.ConversationParticipantDetails;
+import com.google.codeu.chatme.model.User;
```
Remove import (don't think it's needed)

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/tabs/ConversationsFragment.java:
```diff
@@ -1,21 +1,28 @@
 
     private OnFragmentInteractionListener mListener;
-
+    private AppCompatImageButton btnCreateConversation;
```
Can `AppCompatImageButton` be replaced by `ImageButton`?

blong1996 replied:
done!

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/tabs/ConversationsFragment.java:
```diff
@@ -83,6 +92,31 @@ public void onDetach() {
+    public void openCreateConversationActivity() {
+        Intent mIntent = new Intent(getActivity(), CreateConversationActivity.class);
+        getActivity().startActivity(mIntent);
+        getActivity().finish();
```
Replace lines 115-116 with the following:
```
startActivity(mIntent);
finish();
```

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/create/CreateConversationActivity.java:
```diff
@@ -0,0 +1,41 @@
+
+    }
+
+
```
Consider removing extra spacing in this file (will not be addressed by Reformat Code)

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/adapter/UserListAdapter.java:
```diff
@@ -24,12 +29,15 @@
     private List<User> users = new ArrayList<>();
+    private CreateConversationPresenter createConvoPresenter;
 
     private UserPresenter presenter;
```
Rename to `userPresenter`

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/adapter/UserListAdapter.java:
```diff
@@ -70,18 +79,23 @@ public void setUserList(List<User> users) {
+        implements View.OnClickListener {
         private TextView tvName;
         private CircularImageView civUserPic;
+        private String userID;
```
Consider renaming to `userId` for consistency

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/adapter/UserListAdapter.java:
```diff
@@ -97,5 +111,24 @@ private void setHolderPicture(String picUrl) {
+
+        @Override
+        public void onClick(View view) {
+
```
I think you should create the conversation object here and then send that as an argument to `CreateConversationPresenter.createCoversation` method which will add the object to Firebase.

ymittal replied:
David pointed out in one of the earlier PRs that it's better that we just pass the created object to our presenter, so if and when a field in the model class changes, we don't need to change the signature of presenter methods.

blong1996 replied:
okie dokie

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/presenter/CreateConversationInteractor.java:
```diff
@@ -0,0 +1,19 @@
+     *
+     * @param selectedUserID
+     */
+    public Conversation createConversation(String selectedUserID);
```
Do you think `addConversation` would be more apt?

blong1996 replied:
I was following the naming convention of "create" but it can be changed to "add" no biggie

ymittal replied:
Very much appreciated. It's just that according to one of the comments on `UserListAdapter`, passing an already created conversation object might be somewhat better.

blong1996 committed:
Code Review Updates and Reformatting

-Implement recommended code review changes
-reformat all new and edited files

blong1996 committed:
extra space removed in CreateConversationPresenter

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/adapter/UserListAdapter.java:
```diff
@@ -97,5 +110,28 @@ private void setHolderPicture(String picUrl) {
+        @Override
+        public void onClick(View view) {
+            // Create new conversation object, set Owner and Id, and add participant
+            Conversation conversation = new Conversation(createConversationPresenter.getUserId());
```
Use `FirebaseUtil.getCurrentUserId()` instead

ymittal commented on android/app/src/main/java/com/google/codeu/chatme/view/adapter/UserListAdapter.java:
```diff
@@ -97,5 +110,28 @@ private void setHolderPicture(String picUrl) {
+        public void onClick(View view) {
+            // Create new conversation object, set Owner and Id, and add participant
+            Conversation conversation = new Conversation(createConversationPresenter.getUserId());
+            conversation.setId(createConversationPresenter.newConversationId());
```
This suggests that our database structure would look as follows:
```
+--- {conversationId}
     +--- {conversationId}
     +--- {authorId}
     +--- {participants}
```
We probably don't need the `conversationId` as a field in our database.

blong1996 committed:
- Remove Controller Directory
- Add ConversationsPresenter TODOs: Only load conversations with
non-null lastMessage, & Order conversations based on the timestamp for
lastMessage
- Set conversation Id after adding it to DB in
CreateConversationPresenter
- Simplify adding conversation owner in UserListAdapater.OnClick

ymittal reviewed (commented):
LGTM. Once the `FloatingActionButton` is implemented, we should be good to go.

blong1996 committed:
Toolbar Fixes & Floating Create Conversation Button

- implement standard AppBarOverlay Toolbar on all activities
- Add plus sign drawable  for button
- Add floating button to ConversationFragment

ymittal reviewed (approved):
LGTM! 👍 Fantastic stuff, @blong1996. Congratulations.
